core/src/main/java/org/infinispan/interceptors/impl/CacheLoaderInterceptor.java

```diff
@@ -291,11 +294,22 @@ public Object visitSizeCommand(InvocationContext ctx, SizeCommand command) {
    }
 
    private CompletionStage<Long> trySizeOptimization(long flagBitSet) {
-      if (EnumUtil.containsAny(flagBitSet, FlagBitSets.SKIP_CACHE_LOAD | FlagBitSets.SKIP_SIZE_OPTIMIZATION)) {
+      if (EnumUtil.containsAny(flagBitSet, FlagBitSets.SKIP_CACHE_LOAD | FlagBitSets.SKIP_SIZE_OPTIMIZATION)
+            || cacheConfiguration.persistence().passivation()) {
          return NonBlockingStore.SIZE_UNAVAILABLE_FUTURE;
       }
+
       // Get the size from any shared store that isn't async
-      return persistenceManager.size(SHARED.and(NOT_ASYNC));
+      return persistenceManager.size(SHARED.and(NOT_ASYNC))
+            .thenCompose(size -> {
+               if (size > -1
+                     || !EnumUtil.containsAny(flagBitSet, FlagBitSets.CACHE_MODE_LOCAL)
+                     || !cacheConfiguration.persistence().usingSegmentedStore()) {
```

Comment:
I don't believe this line should be here. This makes more sense when merged together with the other size implementation so we can do the size from the private store optimization only if the store is segmented and the segments are provided. If no segments are provided it can always use the private store at this point (since passivation is checked above)

Reply:
Not sure if was correct. Moved this to `ClusteringInterceptor`.